subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/result/VersionSelectionReasons.java

```diff
@@ -131,6 +131,11 @@ public ComponentSelectionReasonInternal addCause(ComponentSelectionDescriptor de
             return ImmutableList.<ComponentSelectionDescriptor>copyOf(descriptions);
         }
 
+        @Override
+        public boolean isConstrained() {
+            return hasCause(ComponentSelectionCause.CONSTRAINT);
```

Comment:
Shouldn't `isExpected() == true` for a constraint that is constrained but resolves to exactly the constraint? Currently, this won't be the case.

Reply:
Now it will.